File: frontend/src/app/dashboard-services/car-issues.service.ts
import { Injectable, Signal, WritableSignal, signal } from "@angular/core";
import { Observable, map } from "rxjs";
import { environment } from "../../environments/environment.development";
import { HttpClient } from "@angular/common/http";
import { CarIssueDTO } from "../models/car-issue-dto";
import { CreateCarIssueDTO } from "../models/create-car-issue-dto";
import { UpdateCarIssueDTO } from "../models/update-car-issue-dto";

@Injectable()
export class CarIssuesService {
  constructor(private _http: HttpClient) {}

  getCarIssues(carId: string): Observable<any> {
    return this._http.get<CarIssueDTO>(
      `${environment.apiUrl}/cars/${carId}/issues`
    );
  }

  createCarIssue(
    carId: string,
    createCarIssueDTO: CreateCarIssueDTO
  ): Observable<CarIssueDTO> {
    return this._http.post<CarIssueDTO>(
      `${environment.apiUrl}/cars/${carId}/issues`,
      createCarIssueDTO
    );
  }

  removeCarIssue(carId: string, issueId: string): Observable<void> {
    return this._http.delete<void>(
      `${environment.apiUrl}/cars/${carId}/issues/${issueId}`
    );
  }

  updateCarIssue(
    carId: string,
    issueId: string,
    carIssueDTO: UpdateCarIssueDTO
  ) {
    return this._http.put<CarIssueDTO>(
      `${environment.apiUrl}/cars/${carId}/issues/${issueId}`,
      carIssueDTO
    );
  }
}

File: backend/Services/CarIssuesService.cs
using System.Security.Claims;
using Backend.Exceptions;
using Backend.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;

public class CarIssuesService(
    ApplicationDbContext applicationDbContext,
    IAuthorizationService authorizationService,
    UserManager<ApplicationUser> userManager) : ICarIssuesService
{

    private readonly ApplicationDbContext applicationDbContext = applicationDbContext;
    private readonly IAuthorizationService authorizationService = authorizationService;
    private readonly UserManager<ApplicationUser> userManager = userManager;

    private async Task<Car> GetCarAsync(ClaimsPrincipal userPrincipal, Guid carId)
    {
        var car = await applicationDbContext.Cars
            .Include(x => x.Issues)
            .Include(x => x.ApplicationUser)
            .FirstOrDefaultAsync(x => x.Id == carId);
        if (car is null)
        {
            throw new ResourceNotFoundException();
        }
        var authorizationResult = await authorizationService.AuthorizeAsync(userPrincipal, car, "SameOwnerPolicy");
        if (authorizationResult.Succeeded)
        {
            return car;
        }
        throw new UnauthorizedException();
    }
    public async Task<CarIssueDTO> CreateCarIssue(ClaimsPrincipal userPrincipal, Guid carId, CreateCarIssueDTO createCarIssueDTO)
    {
        var carIssue = new CarIssue()
        {
            Title = createCarIssueDTO.Title,
            Description = createCarIssueDTO.Description,
            CreateDateTime = createCarIssueDTO.CreateDateTime
        };
        var car = await GetCarAsync(userPrincipal, carId);
        car.Issues.Add(carIssue);
        await applicationDbContext.SaveChangesAsync();
        return new CarIssueDTO(carIssue);
    }

    public async Task<CarIssueDTO[]> GetCarIssues(ClaimsPrincipal userPrincipal, Guid carId)
    {
        var car = await GetCarAsync(userPrincipal, carId);
        var carIssues = car.Issues.Select(x => new CarIssueDTO(x)).ToArray();
        return carIssues;
    }

    public async Task RemoveCarIssue(ClaimsPrincipal userPrincipal, Guid carId, Guid carIssueId)
    {
        var car = await GetCarAsync(userPrincipal, carId);
        var issue = car.Issues.FirstOrDefault(x => x.Id == carIssueId);
        if (issue is not null)
        {
            applicationDbContext.CarIssues.Remove(issue);
            await applicationDbContext.SaveChangesAsync();
        }
    }

    public async Task<CarIssueDTO> UpdateCarIssue(ClaimsPrincipal userPrincipal, Guid carId, Guid carIssueId, UpdateCarIssueDTO carIssueDTO)
    {
        var car = await GetCarAsync(userPrincipal, carId);
        var issue = car.Issues.FirstOrDefault(x => x.Id == carIssueId);
        if (issue is not null)
        {
            issue.Title = carIssueDTO.Title;
            issue.Description = carIssueDTO.Description;
            issue.CreateDateTime = carIssueDTO.CreateDateTime;
            issue.RepairDateTime = carIssueDTO.RepairDateTime;
            issue.RepairCost = carIssueDTO.RepairCost;
            await applicationDbContext.SaveChangesAsync();
            return new CarIssueDTO(issue);
        }
        else
        {
            throw new ResourceNotFoundException();
        }
    }
}